gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -812,58 +836,95 @@ public void toJson(JsonElement jsonElement, JsonWriter writer) throws JsonIOExce
+   *
+   * <p>If you are creating the {@code Type} from a {@link TypeToken}, prefer using
+   * {@link #fromJson(String, TypeToken)} instead since its return type is based on the
+   * {@code TypeToken} and is therefore more type-safe.
+   *
    * @param <T> the type of the desired object
    * @param json the string from which the object is to be deserialized
-   * @param typeOfT The specific genericized type of src. You can obtain this type by using the
-   * {@link com.google.gson.reflect.TypeToken} class. For example, to get the type for
-   * {@code Collection<Foo>}, you should use:
-   * <pre>
-   * Type typeOfT = new TypeToken&lt;Collection&lt;Foo&gt;&gt;(){}.getType();
-   * </pre>
+   * @param typeOfT The specific genericized type of src
    * @return an object of type T from the string. Returns {@code null} if {@code json} is {@code null}
    * or if {@code json} is empty.
-   * @throws JsonParseException if json is not a valid representation for an object of type typeOfT
-   * @throws JsonSyntaxException if json is not a valid representation for an object of type
+   * @throws JsonSyntaxException if json is not a valid representation for an object of type typeOfT
+   *
+   * @see #fromJson(Reader, Type)
+   * @see #fromJson(String, Class)
+   * @see #fromJson(String, TypeToken)
    */
-  @SuppressWarnings("unchecked")
   public <T> T fromJson(String json, Type typeOfT) throws JsonSyntaxException {
     if (json == null) {
       return null;
     }
     StringReader reader = new StringReader(json);
+    @SuppressWarnings("unchecked")
     T target = (T) fromJson(reader, typeOfT);
     return target;
   }
 
   /**
-   * This method deserializes the Json read from the specified reader into an object of the
+   * This method deserializes the specified JSON into an object of the specified type. This method
+   * is useful if the specified object is a generic type. For non-generic objects, use
+   * {@link #fromJson(String, Class)} instead. If you have the JSON in a {@link Reader} instead of
+   * a String, use {@link #fromJson(Reader, TypeToken)} instead.
+   *
+   * @param <T> the type of the desired object
+   * @param json the string from which the object is to be deserialized
+   * @param typeOfT The specific genericized type of src. You should create an anonymous subclass of
+   * {@code TypeToken} with the specific generic type arguments. For example, to get the type for
+   * {@code Collection<Foo>}, you should use:
+   * <pre>
+   * new TypeToken&lt;Collection&lt;Foo&gt;&gt;(){}
+   * </pre>
+   * @return an object of type T from the string. Returns {@code null} if {@code json} is {@code null}
+   * or if {@code json} is empty.
+   * @throws JsonSyntaxException if json is not a valid representation for an object of the erasure type
+   * of typeOfT
+   *
+   * @see #fromJson(Reader, TypeToken)
+   * @see #fromJson(String, Class)
+   */
+  public <T> T fromJson(String json, TypeToken<T> typeOfT) throws JsonSyntaxException {
```

Comment:
I'm not really sure, but would declaring the `typeOfT` as covariant `TypeToken<? extends T>` have any advantages here?

Reply:
I can't currently think of a case where it would matter. Do you have an example?

Reply:
@Marcono1234 
no, I don't, sorry. Variance is probably my weakest point in the Java language, and I'd like to know whether declaring the type token parameter might be affected by the choice. I can't think of a realistic example either as the returned value of type `T` would work fine with both `<T>` and `<? extends T>` since `TypeToken` "produces" only, not "consumes". It seems that not making it covariant would work fine even if the type token class and this method might need something to "consume". So leaving it merely `<T>` is most likely the best choice. I still feel I'm missing to know something. Nevermind.

Reply:
@lyubomyr-shaydariv strictly speaking one is a bounded wildcard and the other is a type parameter, but that is nit-picking (that I bet you are aware of anyway), for the type system these are pretty distinct things. It seems to me that in this particular case it makes no difference at all, so I'll second your reasonings. As a matter of fact, I like this change, I never understood why it never went in initially.

Reply:
I also don't think there's any reason to use `<? extends T>` here.